wallet-new/src/Cardano/Wallet/Kernel/Transactions.hs

```diff
@@ -0,0 +1,279 @@
+pay :: ActiveWallet
+    -> IO Address
+    -- ^ A computation to generate change 'Address'es.
+    -> (Address -> Either CoinSelHardErr SafeSigner)
+    -> CoinSelectionOptions
+    -> HdAccountId
+    -- ^ The source @(root, account) from where the payment was
+    -- originated
+    -> NonEmpty (Address, Coin)
+    -- ^ The payees
+    -> IO (Either PaymentError Tx)
+pay activeWallet genChangeAddr signAddress opts accountId payees = do
+        retrying retryPolicy shouldRetry $ \rs -> do
+            tx <- newTransaction activeWallet genChangeAddr signAddress opts accountId payees
+            case tx of
+                 Left e      -> return (Left $ PaymentNewTransactionError e)
+                 Right txAux -> do
+                     succeeded <- newPending activeWallet accountId txAux
+                     case succeeded of
+                          Left e   -> do
+                              print (sformat build e)
+                              -- If the next retry would bring us to the
+                              -- end of our allowed retries, we fail with
+                              -- a proper error
+                              retriesLeft <- applyPolicy retryPolicy rs
+                              return . Left $ case retriesLeft of
+                                   Nothing ->
+                                       PaymentSubmissionMaxAttemptsReached
+                                   Just _  ->
+                                       PaymentNewPendingError e
+                          Right () -> return . Right . taTx $ txAux
+    where
+        retryPolicy :: RetryPolicyM IO
+        retryPolicy = constantDelay 5000000 <> limitRetries 6
+
+        -- If this is a hard coin selection error we cannot recover, stop
+        -- retrying. If we get a 'Tx' as output, stop retrying immediately.
+        shouldRetry :: RetryStatus -> Either PaymentError Tx -> IO Bool
+        shouldRetry _ (Right _)                             = return False
+        shouldRetry _ (Left (PaymentNewTransactionError _)) = return False
+        shouldRetry _ _                                     = return True
+
+-- | Creates a new 'TxAux' without submitting it to the network.
+newTransaction :: ActiveWallet
+               -> IO Address
+               -- ^ A computation to generate change 'Address'es.
+               -> (Address -> Either CoinSelHardErr SafeSigner)
+               -- ^ A function to sign each Address
+               -> CoinSelectionOptions
+               -- ^ The options describing how to tune the coin selection.
+               -> HdAccountId
+               -- ^ The source @(root, account) from where the payment was
+               -- originated
+               -> NonEmpty (Address, Coin)
+               -- ^ The payees
+               -> IO (Either NewTransactionError TxAux)
+newTransaction ActiveWallet{..} genChangeAddr signAddress options accountId payees = do
+    snapshot <- getWalletSnapshot walletPassive
```

Comment:
Something we have to be wary of here -- the address generation function lives in `IO`, not just because it generates random addresses but presumably also because it registers those addresses with the DB. This means that we're not _actually_ just using a single snapshot, we're in fact modiying the DB as we go. 

In fact, it wouldn't really be necessary to do this. We could change this so it _doesn't_ actually modify the DB, creating new addresses but not registering them. In that way we are _actually_ using just a single snapshot (could modify the type to be `Snapshot -> IO Address` or whatever) and then register them when it has succeeded. That would address your cocern below and also make the flow of data a bit clearer.

Reply:
Me and @edsko had a chat on the above and we concluded that a better design is to factor out the concern of generating an `Address` from the concern of actually running the coin selection algorithm.

I will refactor the `MkTx` function to actually deal with address creation and the `runCoinSelT` will be completely agnostic to the entire change-address generation business.

Reply:
But why does it still take a `genChangeAddr` (and a signer function fort hat matter?) Shouldn't that just be internal responsibility of the wallet?